homeassistant/components/google_travel_time/sensor.py

```diff
@@ -133,63 +94,89 @@ def convert_time_to_utc(timestr):
+        if CONF_UNITS not in options:
+            options[CONF_UNITS] = hass.config.units.name
 
-        travel_mode = config.get(CONF_TRAVEL_MODE)
-        mode = options.get(CONF_MODE)
-
-        if travel_mode is not None:
+        if CONF_TRAVEL_MODE in new_data:
             wstr = (
                 "Google Travel Time: travel_mode is deprecated, please "
                 "add mode to the options dictionary instead!"
             )
             _LOGGER.warning(wstr)
-            if mode is None:
+            travel_mode = new_data.pop(CONF_TRAVEL_MODE)
+            if CONF_MODE not in options:
                 options[CONF_MODE] = travel_mode
 
-        titled_mode = options.get(CONF_MODE).title()
-        formatted_name = f"{DEFAULT_NAME} - {titled_mode}"
-        name = config.get(CONF_NAME, formatted_name)
-        api_key = config.get(CONF_API_KEY)
-        origin = config.get(CONF_ORIGIN)
-        destination = config.get(CONF_DESTINATION)
+        if CONF_MODE not in options:
+            options[CONF_MODE] = "driving"
 
-        sensor = GoogleTravelTimeSensor(
-            hass, name, api_key, origin, destination, options
+        hass.config_entries.async_update_entry(
+            config_entry, data=new_data, options=options
         )
-        hass.data[DATA_KEY].append(sensor)
 
-        if sensor.valid_api_connection:
-            add_entities_callback([sensor])
+    api_key = config_entry.data[CONF_API_KEY]
+    origin = config_entry.data[CONF_ORIGIN]
+    destination = config_entry.data[CONF_DESTINATION]
+    name = name or f"{DEFAULT_NAME}: {origin} -> {destination}"
+
+    if not await hass.async_add_executor_job(
+        is_valid_config_entry, hass, _LOGGER, api_key, origin, destination
+    ):
+        raise ConfigEntryNotReady
+
+    client = Client(api_key, timeout=10)
+
+    sensor = GoogleTravelTimeSensor(
+        hass, config_entry, name, api_key, origin, destination, client
+    )
+
+    async_add_entities([sensor], False)
 
-    # Wait until start event is sent to load this component.
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, run_setup)
+
+async def async_setup_platform(
+    hass: HomeAssistant, config, add_entities_callback, discovery_info=None
+):
+    """Set up the Google travel time platform."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
+    )
+
+    _LOGGER.warning(
+        "Your Google travel time configuration has been imported into the UI; "
+        "please remove it from configuration.yaml as support for it will be "
+        "removed in a future release"
+    )
 
 
 class GoogleTravelTimeSensor(SensorEntity):
     """Representation of a Google travel time sensor."""
 
-    def __init__(self, hass, name, api_key, origin, destination, options):
+    def __init__(self, hass, config_entry, name, api_key, origin, destination, client):
```

Comment:
Don't pass in `hass`. It will be set on the entity when the entity has been added to home assistant. This means you can access `hass` with `self.hass`, after the entity has been added to home assistant.